File: search-a-2d-matrix-74/Solution.java
/**
 * LeetCode problem # 74, 'Search a 2D Matrix'
 * https://leetcode.com/problems/search-a-2d-matrix
 */

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        for (int[] row : matrix) {
            int res = searchRow(row, target);
            if (res >= 0 && res < row.length && row[res] == target) {
                return true;
            }
        }
        return false;
    }

    private int searchRow(int[] row, int target) {
        int low = -1;
        int high = row.length;

        while (low + 1 != high) {
            int mid = low + (high - low) / 2;

            if (row[mid] < target) {
                low = mid;
            } else {
                high = mid;
            }
        }
        return high;
    }
}

File: search-a-2d-matrix-74/solution.js
/**
 * LeetCode problem # 74, 'Search a 2D Matrix'
 * https://leetcode.com/problems/search-a-2d-matrix
 */

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
const searchMatrix = function (matrix, target) {
  for (const row of matrix) {
    const curr = searchRow(row, target);
    if (curr >= 0 && curr <= row.length && row[curr] === target) {
      return true;
    }
  }
  return false;
};

function searchRow(row, target) {
  let lo = -1;
  let hi = row.length;
  while (lo + 1 !== hi) {
    let mid = lo + ((hi - lo) >> 1);
    if (row[mid] < target) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  return hi;
}
